lib/core/src/Cardano/Wallet/Primitive/AddressDiscovery/Delegation.hs

```diff
@@ -0,0 +1,361 @@
+    -> RewardAccount
+nextKey s = accountAtIx s $ nextKeyIx s
+
+lastActiveIx
+    :: DelegationState k
+    -> Maybe (Index 'Soft 'AddressK)
+lastActiveIx s
+    | nextKeyIx s == minBound = Nothing
+    | otherwise               = Just $ pred $ nextKeyIx s
+
+lastActiveKey
+    :: (SoftDerivation k, ToRewardAccount k)
+    => DelegationState k
+    -> Maybe RewardAccount
+lastActiveKey s = accountAtIx s <$> lastActiveIx s
+
+data PointerUTxO = PointerUTxO TxIn Coin
+    deriving (Generic, Eq, Show)
+    deriving anyclass NFData
+
+--------------------------------------------------------------------------------
+-- Constructors
+--------------------------------------------------------------------------------
+
+-- | Construct the initial delegation state.
+initialDelegationState
+    :: k 'AccountK XPub
+    -> DelegationState k
+initialDelegationState accK = DelegationState accK minBound Nothing False
+
+unsafeDeserializeDelegationState
+    :: k 'AccountK XPub
+    -> Index 'Soft 'AddressK
+    -> Maybe PointerUTxO
+    -> Bool
+    -> DelegationState k
+unsafeDeserializeDelegationState = DelegationState
+
+--------------------------------------------------------------------------------
+-- Chain
+--------------------------------------------------------------------------------
+
+data Tx = Tx
+    { certs :: [Cert]
+    , inputs :: [(TxIn, Coin)]
+    , outputs :: [TxOut]
+    }
+    deriving (Eq, Generic)
+    deriving Show via (Quiet Tx)
+
+-- FIXME: This logic might belong in the DBLayer, in which this really is the
+-- MVar implementation and out of place.
+
+data Cert
+    = RegisterKey RewardAccount
+    | Delegate RewardAccount
+      -- ^ Which pool we're delegating to is here (and for now) irrelevant.
+      -- The main thing is that there exists a witness on-chain for this stake
+      -- key (registration certs don't require witnesses)
+    | DeRegisterKey RewardAccount
+    deriving (Eq, Show, Generic)
+
+--
+--
+-- a portfolio of 3 keys:
+-- [0, 1, 2]
+-- should have a nextKeyIx of 3
+--
+-- portfolio of 0 keys, should have a nextKeyIx of 0
+--
+-- FIXME: toEnum safety
+setPortfolioOf
+    :: (SoftDerivation k, ToRewardAccount k)
+    => DelegationState k
+    -> (k 'AddressK XPub -> Address)
+    -> (RewardAccount -> Bool) -- TODO: Need a Set or Map for the real implementation with LSQ
+    -> Int
+    -> Maybe Tx
+setPortfolioOf s mkAddress isReg n =
+    let s' = s { nextKeyIx = toEnum n }
+        mkTxIn (PointerUTxO i c) = (i, c)
+
+        minUTxOVal = Coin 1 -- FIXME
+        -- TODO: Need to rely on wallet to return as change, if the minUTxOVal
+        -- changes. Not sure if this is the case.
+        txWithCerts [] = Nothing
+        txWithCerts cs = Just $ Tx
+            { certs = cs
+            , inputs = maybeToList (mkTxIn <$> pointer s)
+            , outputs =
+                [ TxOut
+                    (mkAddress $ keyAtIx s $ nextKeyIx s')
+                    (TB.fromCoin minUTxOVal)
+                ]
```

Comment:
Initially I was worried that could lead to multiple interfering histories/pointers, but maybe it's not actually a problem if the pointer can only be created in one transition 🤔 Can sanity check in tests…

Reply:
Tests still pass -> cee3a604eb6f279565d7f42fe7ecb2a90df3b3a1 

Good idea!

Reply:
Because of this, we could maybe re-work `DelegationState` to eliminate impossible states a bit:

I.e. rather than
```haskell
    , nextKeyIx :: !(Index 'Soft 'AddressK)

      -- | By threading a long a "pointer UTxO", we leverage the ledger rules to
      -- ensure transactions cannot be re-ordered in a way that breaks the
      -- rules/assumtions for how stake-keys are managed.
    , pointer :: !(Maybe PointerUTxO)

      -- | For compatibility with single-stake-key wallets, we track whether
      -- stake key 0 is registered or not separately.
      --
      -- See the implementaiton of @applyTx@ for how it is used.
    , isKey0Reg :: Bool
```
something more like:
```haskell
data State
    -- | No active stake keys. The initial state of a new wallet.
    = Zero

    -- | The first stake-key (index 0) is registered and either delegating or
    -- about to be delegating.
    | One

    -- | There is more than one active stake keys
    | More
        !(Index 'Soft 'AddressK) -- nextKeyIx - the ix of the next unused key
        !PointerUTxO -- pointer utxo that need to be spent when changing state.
        !Bool -- Is key 0  still registered?
```
👀

Reply:
Done